src/mpid/ch4/generic/gpu/mpidig_gpu_recv.c

```diff
@@ -0,0 +1,81 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2020 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+#include "mpidimpl.h"
+
+int MPIDIG_gpu_recv(void *buf,
+                    MPI_Aint count,
+                    MPI_Datatype datatype,
+                    int rank,
+                    int tag,
+                    MPIR_Comm * comm,
+                    int context_offset, MPIDI_av_entry_t * av,
+                    MPI_Status * status, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_GPU_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_GPU_RECV);
+
+    buf = MPIDIG_gpu_stage_recv_buf(buf, count, datatype);
+
+    mpi_errno =
+        MPIDI_NM_mpi_recv(buf, count, datatype, rank, tag, comm, context_offset, av, status,
+                          request);
+    MPIR_ERR_CHECK(mpi_errno);
+
```

Comment:
@gcongiu I'm watching on this PR also, one thing I'm not clear is that I didn't see the data movement from intermediate CPU buffer back to original GPU buffer after send/recv completes. Can you point it to me? Or it is just draft version now?

Reply:
The data movement is done when the request is freed (and the buffer unstaged/released):         https://github.com/pmodels/mpich/blob/511e80d193864adf430fbf15eedaeb7ad687ce77/src/mpid/ch4/include/mpidpost.h#L41

Reply:
I see, thanks.